### PR TITLE
feat(common): make CompletionQueueImpl mockable [1]

### PR DESCRIPTION
**Pull request by coryan**
Remove all the template member functions from `CompletionQueueImpl`, so
it can be refactored to an abstract class in a future PR. Simplify the
timers for the fake version of `CompletionQueueImpl`.

Part of the work for #4718

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5036)

### CODE REVIEW

**Comment by coryan** [comment]
The API baseline reported a removed function:

```
completion_queue.h
namespace google::cloud::v1
CompletionQueue::MakeDeadlineTimer ( std::chrono::_V2::system_clock::time_point
deadline )
```

This is a false positive, the function moved to the `.h` file where it became inline and thus *appears* as removed because it is no longer in the `.so` file.

File: google/cloud/internal/completion_queue_impl.cc [hunk 1]
```diff
@@ -69,8 +153,39 @@ void CompletionQueueImpl::CancelAll() {
...
+  if (shutdown_) {
+    lk.unlock();
+    op->Notify(/*ok=*/false);
+    return;
+  }
+  auto ins =
+      pending_ops_.emplace(reinterpret_cast<std::intptr_t>(tag), std::move(op));
+  if (ins.second) {
+    start(tag);
+    lk.unlock();
```
Comment (devbww) [inline google/cloud/internal/completion_queue_impl.cc:184]: Remove?

File: google/cloud/internal/completion_queue_impl.cc [hunk 2]
```diff
@@ -69,8 +153,39 @@ void CompletionQueueImpl::CancelAll() {
...
+    return;
+  }
+  google::cloud::internal::ThrowRuntimeError(
+      "assertion failure: insertion should succeed");
```
Comment (devbww) [inline google/cloud/internal/completion_queue_impl.cc:188]: Perhaps something involving "duplicate tag" would be helpful?

Comment (coryan) [reply google/cloud/internal/completion_queue_impl.cc]: Done.

**Review by devbww — APPROVED** [review]

**Pull request merged**